sphereon-kmp-crypto-kms-azure/src/jsMain/kotlin/com/sphereon/crypto/kms/azure/AzureKeyVaultCryptoProvider.js.kt

```diff
@@ -0,0 +1,147 @@
+        clientSecretCredential = AzureIdentity.ClientSecretCredential(
+            config.tenantId,
+            config.credentialOpts.secretCredentialOpts.clientId,
+            config.credentialOpts.secretCredentialOpts.clientSecret
+        )
+    }
+
+    override suspend fun generateKeyAsync(
+        kmsKeyRef: String?,
+        use: JwkUse?,
+        keyOperations: Array<out KeyOperations>?,
+        alg: SignatureAlgorithm?
+    ): ManagedKeyPair {
+        val signatureAlgorithm = alg ?: SignatureAlgorithm.ECDSA_SHA256
+        if (!isSupportedSignatureAlgorithm(signatureAlgorithm)) {
+            throw IllegalArgumentException("Signature algorithm ${signatureAlgorithm.cryptoAlgorithm.name} is not supported by Azure Key Vault")
+        }
+        val keyName = kmsKeyRef ?: "key-${Uuid.v4String()}"
+        val options: AzureKeyvaultKeys.CreateEcKeyOptions = js("{}").unsafeCast<AzureKeyvaultKeys.CreateEcKeyOptions>().apply {
+            curve = signatureAlgorithm.curve?.jose?.value
+        }
+        val keyVaultKey = keyClient.createEcKey(keyName, options).await()
+        val keyVaultJwk = keyVaultKey.toJwk()
+        val publicCoseKey = CoseJoseKeyMappingService.toCoseKey(keyVaultJwk)
+        val kid = keyVaultKey.key.kid
+        return ManagedKeyPair(
+            kms = getId(),
+            kmsKeyRef = keyVaultKey.name,
+            kid = kid,
+            jose = JoseKeyPair(null, keyVaultJwk),
+            cose = CoseKeyPair(null, publicCoseKey)
+        )
+    }
+
+    private fun String.toSignatureAlgorithm(): String {
+        val algorithmMap = mapOf(
+            "P-256" to "ES256",
+            "P-384" to "ES384",
+            "P-521" to "ES512"
+        )
+        return algorithmMap[this] ?: throw IllegalArgumentException("Unsupported algorithm or curve: $this")
+    }
+
+    override suspend fun createRawSignatureAsync(
+        keyInfo: IKeyInfo<*>,
+        input: ByteArray,
+        requireX5Chain: Boolean
+    ): ByteArray {
+        if (keyInfo.kmsKeyRef === null) {
+            throw IllegalArgumentException("Key reference is required")
+        }
+        val azureKey = keyClient.getKey(keyInfo.kmsKeyRef.toString()).await()
+        val cryptographyClient = AzureKeyvaultKeys.CryptographyClient(azureKey, clientSecretCredential)
+        val signature = cryptographyClient.signData(azureKey.key.crv.toSignatureAlgorithm(), input).await()
+        return signature.result
+    }
+
+    override suspend fun isValidRawSignatureAsync(
+        keyInfo: IKeyInfo<*>,
+        input: ByteArray,
+        signature: ByteArray
+    ): Boolean {
+        val azureKey = keyClient.getKey(keyInfo.kmsKeyRef.toString()).await()
+        val cryptographyClient = AzureKeyvaultKeys.CryptographyClient(azureKey, clientSecretCredential)
+        val verifyResult =
+            cryptographyClient.verifyData(azureKey.key.crv.toSignatureAlgorithm(), input, signature).await()
```

Comment:
Why do you need to call toString on this? The kmsKeyRef should always be a string. Also it is a Sphereon construct, which we will store in a DB. So how does Azure know about, unless we kept them the same? In all cases on a IKeyInfo the value does not neceserally have to be there. So could also be a kid for instance

Reply:
It is passed as argument when creating the key, and used later for retrieval. As its always present in the case of Keyvault, I will assert it with !! to get rid of the error message.

Reply:
Hmzz, I think we should add a TODO. It might always be present in case of Azure. But the calling code doesn't necessarily know it would be a KeyVault instance. In the KeyManagementService we should probably internally translate to a Managed key info.

A Key Info object should either have a kid or a key (public), optionally can have kms and kmsKeyRef. A ResolvedKey must have a Key and a ManagedKey must have kms and kmsKeyRef